### ISSUE TITLE
Qusetion: FP8 Allreduce

### ISSUE THREAD

**Issue by MARD1NO** [Azure/MS-AMP#111]
I have noticed in MS-AMP, you use fp8 datatype in communication. 

In TransformerEngine, when using FP8 operation, it need a scale to adjust result. And I wonder in MS-AMP FP8 Allreduce, does it need a scale to adjust allreduce result?....

**Comment by wkcn** [comment]
Hi @MARD1NO , thanks for your attention to our work!

Yes. When performing FP8 weight gradient all-reduce, the high-precision weight gradient is quantized into an FP8 tensor with a scaling factor. The scaling factor is adjusted according to the minimum value of the maximum absolute values in the data-parallel group. We utilize FP8 AllReduce-SUM on the FP8 tensor and the reciprocal of the scaling factor is divided by the data-parallel degree.

Related Code: https://github.com/Azure/MS-AMP/blob/main/msamp/nn/distributed.py#L185
Paper: Eq. 4-6 in https://arxiv.org/pdf/2310.18313.pdf

**Comment by MARD1NO** [comment]
> Hi @MARD1NO , thanks for your attention to our work!
> 
> Yes. When performing FP8 weight gradient all-reduce, the high-precision weight gradient is quantized into an FP8 tensor with a scaling factor. The scaling factor is adjusted according to the minimum value of the maximum absolute values in the data-parallel group. We utilize FP8 AllReduce-SUM on the FP8 tensor and the reciprocal of the scaling factor is divided by the data-parallel degree.
> 
> Related Code: https://github.com/Azure/MS-AMP/blob/main/msamp/nn/distributed.py#L185 Paper: Eq. 4-6 in https://arxiv.org/pdf/2310.18313.pdf

So much thanks, I will take care of these details.

**Comment by afcruzs** [comment]
@wkcn you wrote "The scaling factor is adjusted according to the minimum value of the maximum absolute values in the data-parallel group. "

However, from the code above it seems that you're computing the maximum and then reducing with MAX, no? so that would be the global maximum of the abs of the gradients:
```
# https://github.com/Azure/MS-AMP/blob/main/msamp/nn/distributed.py#L185
            amaxs = torch.stack([g.abs().max().float() for g in grads])
            scales = torch.stack([meta.scale for meta in metas])

            amaxs.nan_to_num_(nan=torch.inf, posinf=torch.inf)    # convert NAN to INF since reduce ignores NAN
            dist.all_reduce(amaxs, op=dist.ReduceOp.MAX, group=self.process_group)
```

Also, the paper states in section 2.1, just before eq 4, that the global minimum is computed. So I see three things:
- Compute the global minimum (no mention to absolute values), stated in the paper
- Compute the global maximum of absolute values of the gradients, what the code seems to do
- Compute the minimum value of the maximum absolute values - what you have mentioned

Am I misunderstanding or conflating something?

**Comment by wkcn** [comment]
@afcruzs  Thank you for pointing it out!

Let me correct it: ""The scaling factor is adjusted according to the global ~~minimum~~ maximum value of the maximum absolute values in the data-parallel group. ""
The scaling factor `scale` is equal to the maximum FP8 representation value divided by the maximum absolute value `amax`, therefore the scaling factor is the global minimum value as shown in the paper.

> Compute the global maximum of absolute values of the gradients, what the code seems to do

It is right.

**Comment by MARD1NO** [comment]
I have another question is in TensorParallel forward, do MS-AMP still use fp8 allreduce ? or it only used in data parallel backward gradient sum steps?

**Comment by wkcn** [comment]
> I have another question is in TensorParallel forward, do MS-AMP still use fp8 allreduce ? or it only used in data parallel backward gradient sum steps?

In tensor parallel forward, the output of the RowParallelLinear is all reduced with BF16 datatype.
FP8 communication is applied on the input of the ColumnParallelLinear (all gather) as well as the weight gradients of Row and Column ParallelLinears (all reduce).

**Comment by MARD1NO** [comment]
> > I have another question is in TensorParallel forward, do MS-AMP still use fp8 allreduce ? or it only used in data parallel backward gradient sum steps?
> 
> In tensor parallel forward, the output of the RowParallelLinear is all reduced with BF16 datatype. FP8 communication is applied on the input of the ColumnParallelLinear (all gather) as well as the weight gradients of Row and Column ParallelLinears (all reduce).

So much thanks! :D
